### PR TITLE
First working version with code extracted from Node.js 15

### PR DESCRIPTION
**Pull request by piranna**
It also has tests, it's compatible with both Node.js and browser, and have been removed Node.js specific code, but lacks of documentation.

### CODE REVIEW

**Comment by TrySound** [comment]
IMO this should be importable module in the first place and then maybe polyfill.

**Comment by piranna** [comment]
> IMO this should be importable module in the first place and then maybe polyfill.

Can you explain a bit more? `randomUUID.js` file is requirable.

**Comment by TrySound** [comment]
index.js is polyfill. Patching builtin objects should not be default.

**Comment by piranna** [comment]
I can change that, but purposse of this project has been to create a polyfill since begining, that's the reason of the `browser` field. But yes, polyfile filename could be renamed to `node.js`, for example.

**Comment by TrySound** [comment]
Polyfill for node + implementation for browser. The api is not part of any spec and is node specific.

**Comment by piranna** [comment]
> The api is not part of any spec

[It is](https://wicg.github.io/uuid/) :-)

**Comment by TrySound** [comment]
Oh, cool. Thanks! I missed this.

**Review by broofa — CHANGES_REQUESTED** [review]
> Polyfill for node + implementation for browser. The api is not part of any spec and is node specific.

Echoing this sentiment.  This is a polyfill for Node's implementation.  While that is the only implementation currently in the wild, I expect that will change fairly quickly as the WICG work progresses.  Specifically:
* `disableEntropyCache` is unlikely to be part of the spec
* If/when type assertions are specified, I would expect them to be `TypeError`s, not the adhoc error classes seen here.

I'm okay moving forward with this but once there's a formal standard, I would assume this module switches to whatever that is (as a breaking change), with Node either leading or following the way on that I guess.

Also, there should be browser tests as well.  Maybe take a page out of the `uuid` module's approach?  (Kind of a nuisance to do, though.  @ctavan: thoughts?)

**Comment by piranna** [comment]
Sorry for the delay. I've reviewed and fixed your concerns, and made sure tests pass on both Node.js (down to v12) and JsDOM environments, with almost 100% tests coverage. [There's an issue](https://github.com/facebook/jest/issues/10547) with `jest` that prevents to resolve the `browser` field, but it's just only a line of code, so we can control it.

Thoughts? Comments?

**Comment by piranna** [comment]
Any update on this? How can we move it forward?

File: README.md
Comment (broofa) [inline README.md:10]:
```suggestion
Require module to polyfill `crypto.randomUUID()` polyfill (if needed).
```

File: README.md
Comment (broofa) [inline README.md:3]:
```suggestion
Stub project for polyfilling `randomUUID` as proposed for standardization in
```

File: README.md
Comment (broofa) [inline README.md:5]:
```suggestion
Polyfill for the `crypto.randomUUID()` method as proposed in
the [WICG randomUUID specification](https://github.com/WICG/uuid) and recently implemented in
[Node.js v15.6.0](https://github.com/nodejs/node/pull/36729).
```

File: README.md
Comment (broofa) [inline README.md:17]:
```suggestion
```

File: README.md
Comment (broofa) [inline README.md:21]: If ESM version won't be available soon, create an issue for it and link to that issue from here.
```suggestion
NOTE: ESM `import`able version not yet available.  See [this issue](appropriate link) for status.
```

File: README.md
Comment (broofa) [inline README.md:32]: Grammer nits.
```suggestion
Note: The current implementation is based on Node.js `lib/internal/crypto/random.js`, however the intent of this project is to follow the spec, so node-specific details (such as the `disableEntropyCache` option) may or may not survive in future versions.
```

File: __tests__/randomUUID.js
Comment (broofa) [inline __tests__/randomUUID.js:28]: Redundant with `testMatch()` (which will fail if either of these fails, right?) 
```suggestion
```

File: __tests__/randomUUID.js
Comment (broofa) [inline __tests__/randomUUID.js:33]: Simpler to just validate the char directly.
```suggestion
    assert.strictEqual(uuid.substr(14, 1), '4');
```

File: __tests__/randomUUID.js
Comment (broofa) [inline __tests__/randomUUID.js:37]: same same.
```suggestion
    assert.match(uuid.substr(19, 1), /[89ab]/);
```

**Review by broofa — APPROVED** [review]
Some nit-picky changes here, but nothing critical.  Approving to allow this to move forward.

Note: A few things I'd like to see added in the near future, if possible:
* code linting (preferably using `uuid`.  I.e. eslint + uuid's `.eslintrc.json`)
* markdown linting (preferably consistent with `uuid`.  I.e. prettier + uuid's `prettier.config.js`)
* CI (preferably GH Action)

**Comment by broofa** [comment]
@ctavan @TrySound, You guys good with this?  I know it's not as far along as `uuid` in terms of linting and CI testing but we can backfill those things as needed.

Also, I suggest we invite @piranna to uuidjs org once this lands.

File: __tests__/randomUUID.js
Comment (piranna) [reply __tests__/randomUUID.js:33]: The test is extracted from Node.js tests too.

File: __tests__/randomUUID.js
Comment (piranna) [reply __tests__/randomUUID.js]: Yes, they would fail. The same, tests extracted from Node.js code.

**Comment by piranna** [comment]
Added linting fixes based on `uuidjs/uuid` project config files. Also applied all suggestions of @broofa.

**Review by broofa — CHANGES_REQUESTED** [review]
It looks like`eslint-plugin-import`, `eslint-plugin-node`,`eslint-plugin-promise`, and `prettier` all need to be added as `devDependencies`.

(Couldn't get `npm test` to work until I installed these.)

**Comment by broofa** [comment]
Oh, we should probably also go ahead and add `"sideEffects": true` to package.json.  From https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free

> A "side effect" is defined as code that performs a special behavior when imported, other than exposing one or more exports. An example of this are polyfills, which affect the global scope and usually do not provide an export.

This is mostly relevent for ESM imports, which aren't supported here yet, but there's no harm in being explicit about this now.

**Comment by piranna** [comment]
> It looks like`eslint-plugin-import`, `eslint-plugin-node`,`eslint-plugin-promise`, and `prettier` all need to be added as `devDependencies`.
> 
> (Couldn't get `npm test` to work until I installed these.)

In my machine I've not needed to add them...

Ubuntu Mate 20.10
Node.js v15.11.0
npm 7.6.0

**Comment by piranna** [comment]
> we should probably also go ahead and add `"sideEffects": true` to package.json

Seems to be really interesting :-)

**Comment by broofa** [comment]
> In my machine I've not needed to add them...

Do you have them installed globally?

**Comment by piranna** [comment]
> Do you have them installed globally?

No, my globals are only these ones:

```
[piranna@PRO2:~/github/randomUUID]
 (main) > npm ls -g
/usr/lib
├── npm-check-updates@10.2.2
├── npm@7.6.0
├── react-native-cli@2.0.1
└── sloc@0.2.1
```

**Comment by broofa** [comment]
> No, my globals are only these ones:

Here's my recipe for recreating the missing modules issue...

```
$ node --version
v14.15.4

$ npm --version
6.14.10

$ rm -fr ./node_modules
...
$ npm install
...
$ npm test
...
Oops! Something went wrong! :(

ESLint: 7.22.0

ESLint couldn't find the plugin "eslint-plugin-import".
```

etc...

**Comment by piranna** [comment]
I think npm v7 is installing missing `peerDependencies`... could that be the reason?

**Comment by piranna** [comment]
Ok, I've checked it with Node.js v14.16.0 and npm v6.14.11 and I'm seeing the same problem as you, so seems it's the thing about npm 7 installing peerDependencies. I'm going to add the missing dependencies.

**Comment by piranna** [comment]
I've added the missing dependencies.

**Review by broofa — APPROVED** [review]

**Comment by broofa** [comment]
My apologies for the delay.  I've gone ahead and landed this since, as previously noted, I think we're at a place where we can push this out and deal with issues as/how they get reported.

I've also taken the following actions (@ctavan, take note):
* Invited @piranna to the `uuidjs` org as a "member"
* `npm owner add piranna` (... but piranna is not appearing yet under `npm owner ls`.  Assuming that's because he needs to accept the invitation?)
* Added randomUUID to the `uuid` team (w/in this org) 
* Enabled the new github "discussions" feature for the `uuid` team

@piranna:  I'll leave it to you to do the `npm publish` honors.

**Comment by broofa** [comment]
Also, thanks for your work and, most of all, your patience.  :-)


image

**Comment by piranna** [comment]
> My apologies for the delay.

No worries :-)


> * Invited @piranna to the `uuidjs` org as a "member"

Thank you so much! :-D


> * `npm owner add piranna` (... but piranna is not appearing yet under `npm owner ls`.  Assuming that's because he needs to accept the invitation?)

I've not received any invitation, just only the one for Github org... Maybe there's some missing step? On https://www.npmjs.com/org/uuid it shows nothing too, but I appear in https://www.npmjs.com/package/randomuuid.


> @piranna: I'll leave it to you to do the `npm publish` honors.

```
403 Forbidden - PUT https://registry.npmjs.org/randomuuid - You cannot publish over the previously published versions: 0.1.0-alpha.0.
```

What version should I use? `0.1.0`?

**Comment by piranna** [comment]
> Also, thanks for your work and, most of all, your patience. :-)

You are welcome, I'm glad to help here :-)

**Comment by broofa** [comment]
> > 403 Forbidden - PUT https://registry.npmjs.org/randomuuid - You cannot publish over the previously published versions: 0.1.0-alpha.0.

> What version should I use? 0.1.0?

I'd go with `1.0.0`.   Minor-bump it once the ESM support gets rolled in.

**Comment by piranna** [comment]
And... [published](https://www.npmjs.com/package/randomuuid) :-)

**Comment by ctavan** [comment]
Woah, amazing work @piranna (and @broofa @TrySound)!

I'm sorry for the radio silence from my end. Started a new job recently so my priorities have shifted a bit. I'm confident that I'll be able to contribute again once I've onboarded :).

**Comment by piranna** [comment]
You are welcome :-)

**Pull request merged**
